### PR TITLE
KAFKA-16965: Throw cause of TimeoutException

### PR DESCRIPTION
**Pull request by aliehsaeedii**
This PR adds the cause of `TimeoutException` for `Producer`.

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java
Comment (mjsax) [inline clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: I think we don't want to throw this exception directly, but rather still throw `TimeoutException` but with a nested `UnknownTopicOrPartitionException` (ie, as "cause" of the timeout)?

Comment (aliehsaeedii) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: > I think we don't want to throw this exception directly, but rather still throw `TimeoutException` but with a nested `UnknownTopicOrPartitionException` (ie, as "cause" of the timeout)?

OK. So I changed the PR title as well:)

**Review by mjsax — APPROVED** [review]
@lianetm @jolshan -- call for a second pair of eyes on this PR.

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:1182]: I may be missing something, but won't this always be false in practice? (and we'll never be able to get a root cause here), or maybe even errors from older responses? This is my reasoning: if we got TimeoutException while awaiting a metadata update, couldn't it be simply that the metadata version did not get bumped (`metadata.update` never happened), therefore, we never made it to the point where we get the errors from the response [here](https://github.com/apache/kafka/blob/a27a458d03a173691053c21ebcfb21f11d3cad12/clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerMetadata.java#L135). Makes sense? 

Honestly, at this point I don't see we're able to get any cause for the timeout in the case that we haven't received a response for the MetadataRequest in the time we have left, which is what should trigger the update [here](https://github.com/apache/kafka/blob/2b491f9611cb8514687b4030f3da95dc7573e1da/clients/src/main/java/org/apache/kafka/clients/NetworkClient.java#L1220)). Are we saying we want to propagate the last known error as cause?

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: Here the metadata update did happen (so we may have received errors in the response), but still, it took longer than our `maxWaitMs` so we want to throw a timeout instead. So in this case, do we really want to show an InvalidTopic/UnauthorizedTopic/... error as the root cause for the timeout?  (Sounds a bit confusing to me as "cause" when the error was not retriable right?, but I totally may be missing how metadata errors like these made us take longer and timeout).  If we had received the response in time, those errors would be propagated in the following `maybeThrowExceptionForTopic`, but here we choose timeout over the errors right? I do get the point on retriables though

**Comment by lianetm** [comment]
Thanks for the changes @aliehsaeedii! Left some comments (I may be missing details about the specific errors that motivated this. Was it mainly trying to know when retrying on retriables like UnknownTopic led to timeout?). Thanks!

**Comment by aliehsaeedii** [comment]
> Was it mainly trying to know when retrying on retriables like UnknownTopic led to timeout?

yes, it was:)

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java
Comment (aliehsaeedii) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:1182]: Thanks. It makes sense! The `if` clause always returns `false`. Thus, it must be removed.

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java
Comment (aliehsaeedii) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: Thanks for the clarification, @lianetm. So, the following cases may happen:
1) a timeout and a non-retriable error (such as InvalidTopic/UnauthorizedTopic) happen
2) a timeout and a retriable error (such as UnknownTopicOrPartition) happen
3) only a non-retriable error happens
- Your worries are about case (1), where non-retriable errors are listed as the cause of the timeout exception, which is itself retriable.
- The purpose of this PR, is to actually know the cause of the timeout exceptions in case (2) to prevent complications such as the one explained in [KAFKA-16508](https://issues.apache.org/jira/browse/KAFKA-16508)
- Case (3) is managed by `maybeThrowExceptionForTopic` and I think we both have no worries about that.

Please correct me if I am wrong or incomplete in listing the cases. If I'm right, changing the code in a way that just retriable errors get listed as the cause of timeout exception ([here](https://github.com/aliehsaeedii/kafka/blob/a27a458d03a173691053c21ebcfb21f11d3cad12/clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java#L1195)) would be easy.

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java
Comment (jolshan) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:1182]: Did we address this comment?

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java
Comment (jolshan) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: Would we see case 1? A timeout and a non-retriable error? If we receive a non-retriable error we don't care about timeouts anymore right?

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java
Comment (jolshan) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:1182]: Also to be clear, there are two types of timeouts here 
1 -- we never get a response back
2 -- we are getting retriable errors over and over (ie unknown topic or partition) until the timeout.

Case 2 is what we hope to improve.

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java
Comment (lianetm) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: @jolshan agree we don't care about timeout if we got a fatal error, but with the changes I was concerned we could get case 1: edge case where we kept retrying, and on the last try before timing out, we got a non-retriable. In that case, our logic here would throw a timeout, but including the the non-retriable as cause right?

Other than this edge case, agree we shouldn't get 1, fatal errors would get propagated on the `maybeThrowExceptionForTopic` a few lines down, all good. 

So seems we are all aligned on reducing the scope to include the "last known error" as cause of the timeout here, but only if it's a retriable one. Sounds good to me to solve what we wanted without messing up the flow for non-retriables, that should remain unchanged I expect.

Comment (jolshan) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: I see. I guess I wasn't sure if we would get to this flow if we timed out with a non-retriable error. (Do we still have this timer if we hit a non-retriable error)

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: what about `assertInstanceOf(UnknownTopicOrPartitionException.class, throwable.getCause());`. Seems clearer and safer (test would fail instead of throwing NPE if cause ends up being null by mistake)

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java
Comment (lianetm) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: yes, it's tricky but we do have it mainly because the flow checks elapsed time before throwing fatals (not the opposite), so we could have an expired "timer" with a fatal in hand (that we haven't thrown just yet).

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java
Comment (lianetm) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:1182]: Good point, I had 1 in mind when not seeing a cause to add here, but defitenly 2 may happen if we are retrying continuously and time runs out while awaitingUpdate (and not on the following elapsed check). So we would want to throw with the cause if errors not null as you initially had @aliehsaeedii , makes sense to me.  

Note that here I don't think we need any check on retriables I would say, because if metadata.getError had had any fatal, it would have been thrown before hitting this line (because the `maybeThrowExceptionForTopic` happens right before this, in the previous iteration, and timeout here means no other metadata update happened). Makes sense?

Comment (jolshan) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:1182]: Yes, makes sense to me.

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java
Comment (jolshan) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: Interesting. I wonder if that is that is the only way. But we don't need to change now I suppose.

Comment (mjsax) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: Interesting -- I would also think that checking for fatal error should be done first (if possible) -- but we should not block this PR on this question either -- maybe, if we file a ticket for updating the log, we could point out the changes in this PR and simplify the code we add here, if this corner case goes away.

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java
Comment (artemlivshits) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:1182]: I have a couple questions:
1. Looks like we're just changing the TimeoutException handling on this codepath, but wouldn't we want to change this across the board, i.e. whenever we generate TimeoutException we actually add the last retriable error (unless it's a true timeout and we simply didn't get a reply)?
2. Rather than do this error state tunnelling across multiple levels, wouldn't it be simpler to just add the root cause whenever we decide to initially throw TimeoutException (i.e. when we get a retriable error but throw TimeoutException instead)?

Also I'm not sure if we get an error in the metadata request if the topic is just missing -- looks the metadata requests would be successful, just not contain the topic.

Comment (lianetm) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:1182]: >Also I'm not sure if we get an error in the metadata request if the topic is just missing -- looks the metadata requests would be successful, just not contain the topic.

Even for missing topics I expect we'll have the topic in the response (with an error code for UNKNOWN_TOPIC_OR_PARTITION). The logic is in how the broker handles the metadata request [here](https://github.com/apache/kafka/blob/645eefe95aa764fdcd1c6390bb2a2e8dc14cde84/core/src/main/scala/kafka/server/KafkaApis.scala#L1274-L1291) I believe.

Comment (aliehsaeedii) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:1181]: > Also I'm not sure if we get an error in the metadata request if the topic is just missing -- looks the metadata requests would be successful, just not contain the topic.

Yes, this was something I raised internally as well. I think if the error message is confessing that the topic is missing, it is missing even with no error.

Comment (lianetm) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:1182]: I will let @aliehsaeedii jump in here, but my 2 cents regarding @artemlivshits point 1 and 2. I totally agree with the view on both, but seems to me that's what we're getting here already?? Trick is that we have different timeouts mixed up here, but let's clarify the concepts: the timeout we want to add a cause to is a "global timeout on retries" (not timeout on a single awaitUpdate). With this, we are indeed :

1. "whenever we generate TimeoutException we actually add the last retriable error" -> we add the cause if we have a last retriable, on the 2 points where a new "global timeout on retries" is created
2. "add the root cause whenever we decide to initially throw TimeoutException" -> similar as above, on the 2 points where we check if the global timeout is expired, we create the exception with the cause (if we have one from a previous retry)

Makes sense?

**Review by lianetm — APPROVED** [review]
Thanks for the changes @aliehsaeedii! Let's make sure @artemlivshits concerns are clarified. I left my thoughts on why I believe we're on the same page, so this LGTM.

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java
Comment (aliehsaeedii) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:1182]: > wouldn't it be simpler to just add the root cause whenever we decide to initially throw TimeoutException

I think 1 and 2 both point to the same thing. So my answer to the main idea:
The 1st timeout exception in producer `waitOnMetadata` originated from [here](https://github.com/apache/kafka/blob/645eefe95aa764fdcd1c6390bb2a2e8dc14cde84/clients/src/main/java/org/apache/kafka/common/utils/SystemTime.java#L58) and we decide on throwing the 2nd one if `elapsed >= maxWaitMs`. Therefore, to have the root cause, we needed kinda tunnelling across multiple levels.

Comment (artemlivshits) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:1182]: > , but seems to me that's what we're getting here already??

I presume there are other timeouts on other code paths, not just in getting metadata?  E.g. when we produce to a topic and the leader has stale metadata, so the client sees the topic present, but the leader hasn't caught up on latest metadata and says that it doesn't know about this topic.  commitTransaction is another codepath that returns timeouts because it waits for sends to complete and sends may be failing with retriable error.  Etc. 

> the timeout we want to add a cause to is a "global timeout on retries" (not timeout on a single awaitUpdate)

I understand that, the intent is to translate the timeout message, not generate a new one, so it can be structured as

```
  String errorMessage = /* translated message with total timeout */
  throw TimeoutException(errorMessage, ex.getCause());
```

Then we can still follow the rule "whoever is creating TimeoutException needs to set the cause" and pass this info more locally without the need to save and tunnel errors around.

Comment (mjsax) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:1182]: That's a good point. I would prefer to do some follow up PR to unify all this code and clean it up nicely. I can file a JIRA for tracking and we should tackle it as immediate follow up before 3.9 goes out.

Comment (artemlivshits) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:1182]: sounds good

**Review by artemlivshits — APPROVED** [review]
LGTM

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java
Comment (mjsax) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:1182]: Thanks. Filed https://issues.apache.org/jira/browse/KAFKA-17019

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java
Comment (jolshan) [inline clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: nit: this comment doesn't seem to apply here and should be above where the error message is defined.
Also just to be clear -- this code path is not one where we see anyting but retriable exceptions? I noticed the other path checks that but not this one.

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java
Comment (jolshan) [inline clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java:956]: nit: do we have a test for all code paths in this PR?

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java
Comment (aliehsaeedii) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: > Also just to be clear -- this code path is not one where we see anyting but retriable exceptions? I noticed the other path checks that but not this one.

Please see https://github.com/apache/kafka/pull/16344#discussion_r1648061530
I personally think that fatal exceptions are thrown [here](https://github.com/apache/kafka/blob/3a0c83e49030074271dd34c20fa99baad6270c93/clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerMetadata.java#L123).

**Review by jolshan — APPROVED** [review]

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:1195]: Excuse me, why add the extra error message exclusively for `RetriableException`? what are the drawbacks if we include the topic exception every time?

Comment (jolshan) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:1195]: Just to make sure I understand correctly--you are asking why we aren't including the throwable for non-retriable exceptions? 

My understanding -- and from the discussion above is that we want clearer errors when we keep retrying over and over until we timeout. I'm not sure if there is a case where we hit this and the error is not retriable?

Comment (chia7712) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:1195]: @jolshan thanks for your response. I understand now.

> I'm not sure if there is a case where we hit this and the error is not retriable?

I have not observed that case either.

**Pull request merged**
